egui_plot/src/items/heatmap.rs

```diff
@@ -0,0 +1,469 @@
+use std::ops::RangeInclusive;
+
+use egui::{Color32, Mesh, NumExt as _, Pos2, Rect, Rgba, Shape, TextStyle, Ui, Vec2, WidgetText};
+
+use crate::{ClosestElem, PlotConfig, PlotGeometry, PlotItem, PlotItemBase, PlotPoint};
+
+use super::{Cursor, LabelFormatter, PlotBounds, PlotTransform};
+use emath::Float as _;
+
+/// Default base colors for heatmap palette
+pub const BASE_COLORS: [Color32; 10] = [
+    Color32::from_rgb(48, 18, 59),
+    Color32::from_rgb(35, 106, 141),
+    Color32::from_rgb(30, 160, 140),
+    Color32::from_rgb(88, 200, 98),
+    Color32::from_rgb(164, 223, 39),
+    Color32::from_rgb(228, 223, 14),
+    Color32::from_rgb(250, 187, 13),
+    Color32::from_rgb(246, 135, 8),
+    Color32::from_rgb(213, 68, 2),
+    Color32::from_rgb(122, 4, 2),
+];
+
+#[derive(Debug)]
+pub enum HeatmapErr {
+    ZeroColumns,
+    ZeroRows,
+    BadLength,
+}
+
+impl std::fmt::Display for HeatmapErr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::ZeroColumns => write!(f, "number of columns must not be zero"),
+            Self::ZeroRows => write!(f, "number of rows must not be zero"),
+            Self::BadLength => write!(
+                f,
+                "length of value vector is not divisible by number of columns"
+            ),
+        }
+    }
+}
+
+impl std::error::Error for HeatmapErr {}
+
+/// A heatmap.
+pub struct Heatmap<const RESOLUTION: usize> {
+    base: PlotItemBase,
+
+    /// occupied space in absolute plot coordinates
```

Comment:
Nit: Please use proper English sentences, i.e. `Occupied space in absolute plot coordinates.`